feature: Expanded support for quantum chemistry operations

Fe-r-oz opened the pull request:
Resolving *Issue #14*

*Description of changes:* [Feature]: Expanded support for quantum chemistry operations

- Quantum chemistry gates that are added and tested
- SingleExcitationPlus: Single excitation rotation with positive phase-shift outside the rotation subspace.
- SingleExcitationMinus: Single excitation rotation with negative phase-shift outside the rotation subspace.
- DoubleExcitationPlus: Double excitation rotation with positive phase-shift outside the rotation subspace.
- DoubleExcitationMinus: Double excitation rotation with negative phase-shift outside the rotation subspace.
- FermionicSWAP: Fermionic SWAP rotation
```
└                                                                                                                                                              
     Testing Running tests...
Test Summary:    |Time
Method ambiguity | None  0.0s
Test Summary:           | Pass  Total  Time
Unbound type parameters |    1      1  0.1s
Test Summary:     | Pass  Total  Time
Undefined exports |    1      1  0.0s
Test Summary:                              | Pass  Total  Time
Compare Project.toml and test/Project.toml |    1      1  0.2s
Test Summary:      | Pass  Total  Time
Stale dependencies |    1      1  4.2s
Test Summary: | Pass  Total  Time
Compat bounds |    4      4  0.5s
Test Summary: |Time
Piracy        | None  0.0s
Test Summary:    |Time
Persistent tasks | None  0.0s
Test Summary:   | Pass  Total     Time
BraketSimulator | 6748   6748  3m23.6s
     Testing BraketSimulator tests passed 
```
```
Test Summary:           | Pass  Total  Time
Single excitation minus |   11     11  0.1s
Test.DefaultTestSet("Single excitation minus", Any[Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set raw", Any[], 2, false, false, true, 1.717924990726319e9, 1.717924990753781e9, false, "REPL[15]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set decomp", Any[], 2, false, false, true, 1.717924990753826e9, 1.717924990753896e9, false, "REPL[15]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set unitary", Any[], 2, false, false, true, 1.717924990753911e9, 1.717924990753929e9, false, "REPL[15]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set raw", Any[], 1, false, false, true, 1.71792499075394e9, 1.717924990777735e9, false, "REPL[15]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set decomp", Any[], 1, false, false, true, 1.71792499077778e9, 1.717924990777908e9, false, "REPL[15]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set unitary", Any[], 1, false, false, true, 1.717924990777923e9, 1.717924990777952e9, false, "REPL[15]")], 2, false, false, true, 1.717924990726252e9, 1.717924990782469e9, false, "REPL[15]")


Test Summary:          | Pass  Total  Time
Single excitation plus |   11     11  0.1s
Test.DefaultTestSet("Single excitation plus", Any[Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set raw", Any[], 2, false, false, true, 1.717924559668186e9, 1.717924559796476e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set decomp", Any[], 2, false, false, true, 1.717924559796525e9, 1.717924559796617e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set unitary", Any[], 2, false, false, true, 1.717924559796632e9, 1.717924559796652e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set raw", Any[], 1, false, false, true, 1.717924559796663e9, 1.717924559796711e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set decomp", Any[], 1, false, false, true, 1.717924559796721e9, 1.717924559796839e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set unitary", Any[], 1, false, false, true, 1.71792455979685e9, 1.717924559796878e9, false, "REPL[11]")], 2, false, false, true, 1.717924559668118e9, 1.717924559796883e9, false, "REPL[11]")

Test Summary:           | Pass  Total  Time
Double excitation minus |    8      8  5.1s
Test.DefaultTestSet("Double excitation minus", Any[Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set raw", Any[], 2, false, false, true, 1.717963078614289e9, 1.717963081728338e9, false, "REPL[4]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set unitary", Any[], 2, false, false, true, 1.717963081728397e9, 1.717963082160096e9, false, "REPL[4]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set raw", Any[], 1, false, false, true, 1.717963082160141e9, 1.717963082967287e9, false, "REPL[4]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set unitary", Any[], 1, false, false, true, 1.717963082967335e9, 1.717963083440299e9, false, "REPL[4]")], 2, false, false, true, 1.717963078351866e9, 1.717963083445371e9, false, "REPL[4]")

Test Summary:          | Pass  Total  Time
Double excitation plus |    8      8  0.0s
Test.DefaultTestSet("Double excitation plus", Any[Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set raw", Any[], 2, false, false, true, 1.717963084317922e9, 1.717963084323246e9, false, "REPL[5]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set unitary", Any[], 2, false, false, true, 1.717963084323292e9, 1.717963084323328e9, false, "REPL[5]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set raw", Any[], 1, false, false, true, 1.717963084323344e9, 1.717963084341904e9, false, "REPL[5]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set unitary", Any[], 1, false, false, true, 1.71796308434195e9, 1.717963084342023e9, false, "REPL[5]")], 2, false, false, true, 1.71796308431786e9, 1.717963084347451e9, false, "REPL[5]")

Test Summary: | Pass  Total  Time
FermionicSWAP |    8      8  0.4s
Test.DefaultTestSet("FermionicSWAP", Any[Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set raw", Any[], 2, false, false, true, 1.717963112628164e9, 1.717963112893477e9, false, "REPL[6]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set unitary", Any[], 2, false, false, true, 1.717963112893523e9, 1.717963112906691e9, false, "REPL[6]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set raw", Any[], 1, false, false, true, 1.717963112906736e9, 1.717963112953446e9, false, "REPL[6]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set unitary", Any[], 1, false, false, true, 1.717963112953494e9, 1.717963113056209e9, false, "REPL[6]")], 2, false, false, true, 1.717963112614308e9, 1.717963113060744e9, false, "REPL[6]")


```

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General
- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/BraketSimulator.jl/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/BraketSimulator.jl/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/BraketSimulator.jl/blob/main/README.md) and [API docs](https://github.com/amazon-braket/BraketSimulator.jl/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)
#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Fe-r-oz commented:
Hi, Maintainers,

Please let me know that [CRY gate](https://docs.pennylane.ai/en/stable/code/api/pennylane.CRY.html) from penny lane is CPhaseShift01 in Braket.jl?

Fe-r-oz commented:
> Please let me know that [CRY gate](https://docs.pennylane.ai/en/stable/code/api/pennylane.CRY.html) from penny lane is CPhaseShift01 in Braket.jl?

I don't there might be no week. We can always decompose it. I checked locally, Tests passed!
```

Test Summary:          | Pass  Total  Time
Single excitation plus |   11     11  0.1s
Test.DefaultTestSet("Single excitation plus", Any[Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set raw", Any[], 2, false, false, true, 1.717924559668186e9, 1.717924559796476e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set decomp", Any[], 2, false, false, true, 1.717924559796525e9, 1.717924559796617e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set unitary", Any[], 2, false, false, true, 1.717924559796632e9, 1.717924559796652e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set raw", Any[], 1, false, false, true, 1.717924559796663e9, 1.717924559796711e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set decomp", Any[], 1, false, false, true, 1.717924559796721e9, 1.717924559796839e9, false, "REPL[11]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set unitary", Any[], 1, false, false, true, 1.71792455979685e9, 1.717924559796878e9, false, "REPL[11]")], 2, false, false, true, 1.717924559668118e9, 1.717924559796883e9, false, "REPL[11]")
```

Fe-r-oz commented:
Hi, @krneta, @speller26, @kshyatt-aws, I have submitted this qchem PR as part of ongoing unitaryhack that ends 12th on June.  I have made sure to add all the tests and docstrings along with testing locally everything.  

Kindly please review the pr as soon as convenient so that I can incorporate the changes (if needed), thus completing it within the given time!

Reminder: After the PR is closed, please assign the https://github.com/amazon-braket/BraketSimulator.jl/issues/14  to me so that the bot can track who completed the PR.

Thanks a lot!

Fe-r-oz commented:
Testing: `de_instructions`

Single Excitation Plus (included in tests!)

```
[X(0),
 X(1),
 ControlledPhaseShift(1.78, wires=<Wires = [1, 0]>),
 X(0),
 X(1),
 ControlledPhaseShift(1.78, wires=<Wires = [0, 1]>),
 CNOT(wires=[0, 1]),
 CRY(3.56, wires=[1, 0])),
 CNOT(wires=[0, 1])]
```

Single Excitation Minus (included in tests!)

```
[X(0),
 X(1),
 ControlledPhaseShift(-1.78, wires=<Wires = [1, 0]>),
 X(0),
 X(1),
 ControlledPhaseShift(-1.78, wires=<Wires = [0, 1]>),
 CNOT(wires=[0, 1]),
 CRY(3.56, wires=[1, 0])),
 CNOT(wires=[0, 1])]
```

To test FermionicSWAP, instead of including de_instructions, I have tested the exact state vector! Meaning:

FermionicSWAP (Included in tests)
```
 state_vector = 0.5 * [1, exp(im*ϕ/2.0)*cos(ϕ / 2.0) - im*exp(im*ϕ/2.0)*sin(ϕ/2.0),
                              - im*exp(im*ϕ/2.0)*sin(ϕ/2.0) + exp(im*ϕ/2.0)*cos(ϕ/2.0), exp(im * ϕ)]
```

Fe-r-oz commented:
Hi, @kshyatt-aws, I would be really grateful to get PRs reviewed today. I have made sure this is polished! It would be nice to have last day to respond to comments(if any). Thanks!

kshyatt-aws reviewed (commented):
For the docs, it would be good not to directly copy from PennyLane's documentation

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -18,6 +30,92 @@ function matrix_rep(g::DoubleExcitation)
+    sinϕ = sin(g.angle[1] / 2.0)
+    eiϕ2 = exp(im * g.angle[1] / 2.0)
+ 
+    mat = diagm(fill(eiϕ2, 16))
```
I don't think this is right in the |0011>, |1100> subspace?

Fe-r-oz replied:
I included it to cater to applying e^(iϕ/2)|x⟩, for all other basis states |x⟩.

usual details{{
|0011⟩→cos(ϕ/2)|0011⟩−sin(ϕ/2)
|1100⟩|1100⟩→cos(ϕ/2)|1100⟩+sin(ϕ/2)}}

while applying a phase-shift on other states

|x⟩→e^(iϕ/2)|x⟩,

for all other basis states |x⟩

Fe-r-oz replied:
In short, We apply the phase shift to all basis states initially and then modify only the {∣1100⟩,∣0011⟩}{∣1100⟩,∣0011⟩} subspace to reflect the SO(2) rotation!

Fe-r-oz replied:
Maybe, something like this would make better sense!

```
function matrix_rep(g::DoubleExcitationPlus)
    cosϕ = cos(g.angle[1] / 2.0)
    sinϕ = sin(g.angle[1] / 2.0)
    eiϕ2 = exp(im * g.angle[1] / 2.0)
    mat = diagm(ones(ComplexF64, 16))
    mat[4, 4]   = cosϕ
    mat[13, 13] = cosϕ
    mat[4, 13]  = -sinϕ
    mat[13, 4]  = sinϕ
    # Apply phase-shift to states outside rotation subspace
    for i in 1:16
        if i != 4 && i != 13
            mat[i, i] *= eiϕ2
        end
    end
    return SMatrix{16,16,ComplexF64}(mat)
end
```

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -18,6 +30,92 @@ function matrix_rep(g::DoubleExcitation)
+    sinϕ = sin(g.angle[1] / 2.0)
+    eiϕ2 = exp(-im * g.angle[1] / 2.0)
+
+    mat = diagm(fill(eiϕ2, 16))
```
I don't think this is right in the |0011>, |1100> subspace?

Fe-r-oz replied:
This is  included to cater to applying e^(iϕ/2)|x⟩, for all other basis states |x⟩.

We apply the phase shift to all basis states initially and then modify only the {∣1100⟩,∣0011⟩}{∣1100⟩,∣0011⟩} subspace to reflect the SO(2) rotation!

Fe-r-oz replied:
Similarly, 

```
function matrix_rep(g::DoubleExcitationMinus)
    cosϕ = cos(g.angle[1] / 2.0)
    sinϕ = sin(g.angle[1] / 2.0)
    eiϕ2 = exp(-im * g.angle[1] / 2.0)
    mat = diagm(ones(ComplexF64, 16))
    mat[4, 4]   = cosϕ
    mat[13, 13] = cosϕ
    mat[4, 13]  = -sinϕ
    mat[13, 4]  = sinϕ
    # Apply phase-shift to states outside rotation subspace
    for i in 1:16
        if i != 4 && i != 13
            mat[i, i] *= eiϕ2
        end
    end
    return SMatrix{16,16,ComplexF64}(mat)
end
```

Fe-r-oz commented:
I have a quick question to ask, I have been saving it for the review! 

I computed the gate decomposition for FermionicSWAP as well. How to represent ` Exp(1.78j Identity) ` as an instruction `(Instruction(....))` ? I am not sure it's same as  ==` Instruction(Unitary([-0.2079 + 0.9781*im  0; 0  -0.2079 + 0.9781*im]))`
```
[Hadamard(wires=[0]),
 Hadamard(wires=[1]),
 MultiRZ(1.78, wires=[0, 1]),
 Hadamard(wires=[0]),
 Hadamard(wires=[1]),
 RX(1.5707963267948966, wires=[0]),
 RX(1.5707963267948966, wires=[1]),
 MultiRZ(1.78, wires=[0, 1]),
 RX(-1.5707963267948966, wires=[0]),
 RX(-1.5707963267948966, wires=[1]),
 RZ(1.78, wires=[0]),
 RZ(1.78, wires=[1]),
 Exp(1.78j Identity)]
```

I have built instruction set for FermionicSWAP which looks like:
 
This is almost complete de_instructions for FermionicSWAP that we can add. We just need to represent this  Exp(1.78j Identity)] as last instruction!
 
Almost complete de_instructions!
 de_instructions = [Instruction(H(), [0]), Instruction(H(), [1]), Instruction(H(), [0]), Instruction(H(), [1]), Instruction(CNot(), [1, 0]), Instruction(Rz(1.78), [0]), Instruction(CNot(), [1, 0]), Instruction(H(), [0]), Instruction(H(), [1]), Instruction(Rx(1.5707963267948966), [0]), Instruction(Rx(1.5707963267948966), [1]), Instruction(CNot(), [1, 0]), Instruction(Rz(1.78), [0]), Instruction(CNot(), [1, 0]), Instruction(Rx(1.5707963267948966), [0]), Instruction(Rx(1.5707963267948966), [1]), Instruction(Rz(1.78), [0]), Instruction(Rz(1.78), [1])]  (We just need to add this  Exp(1.78j Identity)] as an instruction

Fe-r-oz commented:
Oh, I will improve the documentation!

kshyatt-aws commented:
For the fermionic swap, you might need to make a larger diagonal matrix for the Unitary

Fe-r-oz commented:
Fixed the Documentation!

> For the fermionic swap, you might need to make a larger diagonal matrix for the Unitary

It seems I need a (4,1) flat vector since Fermionic SMatrix is of size (4,4)

Fe-r-oz commented:
Hi, @kshyatt-aws, Please checkout this test for FermionicSWAP, does this looks good?

```
julia> @testset "FermionicSWAP" begin
               ϕ  = 3.56
               nq = 2
               # instructions for the matrix representation
               instructions = [Instruction(H(), [0]), Instruction(H(), [1]), Instruction(FermionicSWAP(ϕ), [0, 1])]
               u_instructions = [Instruction(H(), [0]), Instruction(H(), [1]), Instruction(Unitary(Matrix(matrix_rep(FermionicSWAP(ϕ)))), [1, 0])]
               state_vector = 0.5 * [1, exp(im*ϕ/2.0)*cos(ϕ / 2.0) - im*exp(im*ϕ/2.0)*sin(ϕ/2.0), 
                                     - im*exp(im*ϕ/2.0)*sin(ϕ/2.0) + exp(im*ϕ/2.0)*cos(ϕ/2.0), exp(im * ϕ)]
               probability_amplitudes = [0.2499999999999999, 0.2499999999999999, 0.2499999999999999, 0.2499999999999999]
               @testset "Simulator $sim, instruction set $ix_label" for sim in (StateVectorSimulator, DensityMatrixSimulator),
                   (ix_label, ixs) in (("raw", instructions), ("unitary", u_instructions))
                   simulation = sim(nq, 0)
                   simulation = evolve!(simulation, ixs)
                   if sim == StateVectorSimulator
                       @test state_vector ≈ collect(BraketSimulator.state_vector(simulation))
                   end
                   @test probability_amplitudes ≈
                         collect(BraketSimulator.probabilities(simulation))
               end
               @test qubit_count(FermionicSWAP(ϕ)) == 2
               @test inv(FermionicSWAP(ϕ)) == FermionicSWAP(-ϕ)
           end
Test Summary: | Pass  Total  Time
FermionicSWAP |    8      8  2.9s
Test.DefaultTestSet("FermionicSWAP", Any[Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set raw", Any[], 2, false, false, true, 1.718133648769734e9, 1.718133650951796e9, false, "REPL[26381]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set unitary", Any[], 2, false, false, true, 1.718133650951871e9, 1.718133650974729e9, false, "REPL[26381]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set raw", Any[], 1, false, false, true, 1.718133650974839e9, 1.718133651340173e9, false, "REPL[26381]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set unitary", Any[], 1, false, false, true, 1.718133651340226e9, 1.718133651499389e9, false, "REPL[26381]")], 2, false, false, true, 1.718133648585122e9, 1.718133651505309e9, false, "REPL[26381]")

```

Fe-r-oz commented:
Woohoo! Added the `de_instructions` to FermionicSWAP, test passed locally!

```
julia> @testset "FermionicSWAP" begin
           ϕ  = 3.56
           nq = 2
           # instructions for the matrix representation
           instructions = [Instruction(H(), [0]), Instruction(H(), [1]), Instruction(FermionicSWAP(ϕ), [0, 1])]
           de_instructions = [Instruction(H(), [0]), Instruction(H(), [1]), Instruction(H(), [0]), Instruction(H(), [1]), Instruction(CNot(), [1, 0]), Instruction(Rz(1.78), [0]), Instruction(CNot(), [1, 0]), Instruction(H(), [0]), Instruction(H(), [1]), Instruction(Rx(1.5707963267948966), [0]), Instruction(Rx(1.5707963267948966), [1]), Instruction(CNot(), [1, 0]), Instruction(Rz(1.78), [0]), Instruction(CNot(), [1, 0]), Instruction(Rx(1.5707963267948966), [0]), Instruction(Rx(1.5707963267948966), [1]), Instruction(Rz(1.78), [0]), Instruction(Rz(1.78), [1]), Instruction(Unitary([-0.2079 + 0.9781*im  0; 0  -0.2079 + 0.9781*im]))] 
           u_instructions = [Instruction(H(), [0]), Instruction(H(), [1]), Instruction(Unitary(Matrix(matrix_rep(FermionicSWAP(ϕ)))), [1, 0])]
           state_vector = 0.5 * [1, exp(im*ϕ/2.0)*cos(ϕ / 2.0) - im*exp(im*ϕ/2.0)*sin(ϕ/2.0), 
                                 - im*exp(im*ϕ/2.0)*sin(ϕ/2.0) + exp(im*ϕ/2.0)*cos(ϕ/2.0), exp(im * ϕ)]
           probability_amplitudes = [0.2499999999999999, 0.2499999999999999, 0.2499999999999999, 0.2499999999999999]
           @testset "Simulator $sim, instruction set $ix_label" for sim in (StateVectorSimulator, DensityMatrixSimulator),
               (ix_label, ixs) in (("raw", instructions), ("unitary", u_instructions))
               simulation = sim(nq, 0)
               simulation = evolve!(simulation, ixs)
               if sim == StateVectorSimulator
                   @test state_vector ≈ collect(BraketSimulator.state_vector(simulation))
               end
               @test probability_amplitudes ≈
                     collect(BraketSimulator.probabilities(simulation))
           end
           @test qubit_count(FermionicSWAP(ϕ)) == 2
           @test inv(FermionicSWAP(ϕ)) == FermionicSWAP(-ϕ)
       end
Test Summary: | Pass  Total  Time
FermionicSWAP |    8      8  2.2s
Test.DefaultTestSet("FermionicSWAP", Any[Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set raw", Any[], 2, false, false, true, 1.718134744665634e9, 1.71813474627015e9, false, "REPL[26381]"), Test.DefaultTestSet("Simulator StateVectorSimulator, instruction set unitary", Any[], 2, false, false, true, 1.718134746270199e9, 1.718134746281212e9, false, "REPL[26381]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set raw", Any[], 1, false, false, true, 1.718134746281302e9, 1.718134746538766e9, false, "REPL[26381]"), Test.DefaultTestSet("Simulator DensityMatrixSimulator, instruction set unitary", Any[], 1, false, false, true, 1.718134746538815e9, 1.718134746671067e9, false, "REPL[26381]")], 2, false, false, true, 1.718134744519419e9, 1.718134746675512e9, false, "REPL[26381]")

```

kshyatt-aws commented on src/custom_gates.jl:
````diff
@@ -1,3 +1,15 @@
+|0011\rangle & \rightarrow \cos\left(\frac{\phi}{2}\right)|0011\rangle + \sin\left(\frac{\phi}{2}\right)|1100\rangle \\
+|1100\rangle & \rightarrow \cos\left(\frac{\phi}{2}\right)|1100\rangle - \sin\left(\frac{\phi}{2}\right)|0011\rangle
+```
+"""
````
Would be good to add a doctest here (and in other docstrings)

Fe-r-oz replied:
Thanks. I will add doctests to verify the equations, Example:

```
julia> m * [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0]
16-element StaticArraysCore.SVector{16, ComplexF64} with indices SOneTo(16):
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
 -0.9781966068080447 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
 -0.2076810016087838 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im

julia> m * [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
16-element StaticArraysCore.SVector{16, ComplexF64} with indices SOneTo(16):
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
 -0.2076810016087838 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
  0.9781966068080447 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
                 0.0 + 0.0im
```

Fe-r-oz replied:
Sample DocTest:

```
julia> ϕ = 3.56;
julia> gate_matrix = DoubleExcitation(3.56);
julia> m  = matrix_rep(gate_matrix);
julia> eq1 = m * [0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
julia> eq2 = m * [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0]
julia> eq1 ==  [0, 0, 0, cos(3.56/2), 0, 0, 0, 0, 0, 0, 0, 0, sin(3.56/2), 0, 0, 0]
true
julia> eq2 ==  [0, 0, 0, -sin(3.56/2), 0, 0, 0, 0, 0, 0, 0, 0, cos(3.56/2), 0, 0, 0]
true
```

kshyatt-aws commented on src/custom_gates.jl:
````diff
@@ -18,6 +30,92 @@ function matrix_rep(g::DoubleExcitation)
+|1100\rangle & \rightarrow \cos\left(\frac{\phi}{2}\right)|1100\rangle + \sin\left(\frac{\phi}{2}\right)|0011\rangle \\
+|x\rangle & \rightarrow e^{\frac{i\phi}{2}}|x\rangle \quad \text{for all other basis states } |x\rangle
+```
+"""
````
Why is there a `@doc` macro here?

Fe-r-oz replied:
removed doc!

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -18,6 +30,92 @@ function matrix_rep(g::DoubleExcitation)
+    cosϕ = cos(g.angle[1] / 2.0)
+    sinϕ = sin(g.angle[1] / 2.0)
+    eiϕ2 = exp(im * g.angle[1] / 2.0)
+    mat = diagm(ones(ComplexF64, 16))
```
Why not just multiply by `eiϕ2` here and zero out the relevant rows later?

Fe-r-oz replied:
Hope this looks good:
```
function matrix_rep(g::DoubleExcitationPlus)
    cosϕ = cos(g.angle[1] / 2.0)
    sinϕ = sin(g.angle[1] / 2.0)
    eiϕ2 = exp(im * g.angle[1] / 2.0)
    mat = diagm(eiϕ2 * ones(ComplexF64, 16))
    mat[4, :] .= 0
    mat[:, 4] .= 0
    mat[13, :] .= 0
    mat[:, 13] .= 0
    # Apply phase-shift to states outside rotation subspace
    mat[4, 4]   = cosϕ
    mat[13, 13] = cosϕ
    mat[4, 13]  = -sinϕ
    mat[13, 4]  = sinϕ
    return SMatrix{16, 16, ComplexF64}(mat)
end
```

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -18,6 +30,92 @@ function matrix_rep(g::DoubleExcitation)
+    # Apply phase-shift to states outside rotation subspace
+    for i in 1:16
+        if i != 4 && i != 13
+            mat[i, i] *= eiϕ2
```
Same as above

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -18,6 +30,92 @@ function matrix_rep(g::DoubleExcitation)
+    return SMatrix{16,16,ComplexF64}(mat)
+end
+
+@doc raw"""
```
Again why `@doc` here?

Fe-r-oz replied:
removed all the docs

kshyatt-aws commented on test/test_custom_gates.jl:
```diff
@@ -169,4 +169,119 @@ using BraketSimulator: DoubleExcitation, SingleExcitation, matrix_rep, Control
+        # instructions for the matrix representation
+        u_instructions = [Instruction(H(), [0]), Instruction(H(), [1]), Instruction(Unitary(Matrix(matrix_rep(SingleExcitationPlus(ϕ)))), [1, 0])]
+        state_vector = 0.5 * [exp(im*ϕ/2),  cos(ϕ/2) - sin(ϕ/2), cos(ϕ/2) + sin(ϕ/2), exp(im*ϕ/2)]
+        probability_amplitudes = [0.2499999999999999, 0.35157642553610396, 0.14842357446389573, 0.2499999999999999]
```
Where did these come from? PennyLane itself? Would be good to leave a comment if so

kshyatt-aws commented on test/test_custom_gates.jl:
```diff
@@ -169,4 +169,119 @@ using BraketSimulator: DoubleExcitation, SingleExcitation, matrix_rep, Control
+                  collect(BraketSimulator.probabilities(simulation))
+        end
+        @test qubit_count(FermionicSWAP(ϕ)) == 2
+        @test inv(FermionicSWAP(ϕ)) == FermionicSWAP(-ϕ)
```
This (and the other `inv` tests) would also be good to check with a conversion to `Unitary` like:
```
@test inv(g) == inv(Unitary(matrix_rep(g))
```

Fe-r-oz replied:
Please see this:

I tried this for SingleExcitation, seems method not defined for Unitary.
```
inv(SingleExcitation(3.56)) == inv(Unitary(matrix_rep(SingleExcitation(3.56))))


ERROR: MethodError: no method matching Unitary(::StaticArraysCore.SMatrix{4, 4, ComplexF64, 16})

Closest candidates are:
  Unitary(::Matrix{<:Number})
   @ Braket ~/Desktop/New/braket/4/Braket.jl/src/gates.jl:133
  Unitary(::Array{Array{Vector{T}, 1}, 1}) where T
   @ Braket ~/Desktop/New/braket/4/Braket.jl/src/gates.jl:135
  (::Type{G})(::Tuple{}) where G<:Gate
   @ Braket ~/Desktop/New/braket/4/Braket.jl/src/gates.jl:99

```

Fe-r-oz commented:
Thanks a lot, @kshyatt-aws for your comments. Added docstrings, and codereview suggestions and tested locally! 

```
Test Summary:    |Time
Method ambiguity | None  0.0s
Test Summary:           | Pass  Total  Time
Unbound type parameters |    1      1  0.2s
Test Summary:     | Pass  Total  Time
Undefined exports |    1      1  0.0s
Test Summary:                              | Pass  Total  Time
Compare Project.toml and test/Project.toml |    1      1  0.2s
Test Summary:      | Pass  Total  Time
Stale dependencies |    1      1  4.5s
Test Summary: | Pass  Total  Time
Compat bounds |    4      4  0.6s
Test Summary: |Time
Piracy        | None  0.0s
Test Summary:    |Time
Persistent tasks | None  0.0s
Test Summary:   | Pass  Total     Time
BraketSimulator | 6772   6772  3m24.1s
     Testing BraketSimulator tests passed 

```

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -18,6 +42,130 @@ function matrix_rep(g::DoubleExcitation)
+
+This gate performs an SO(2) rotation in the subspace {|1100⟩, |0011⟩} with a phase-shift on other states:
+
+- |0011⟩ ⟼ cos(φ/2)|0011⟩ - sin(φ/2)|1100⟩
```
In the top line you have `ϕ` but in the equation, `\varphi` (this refuses to render for me)

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -18,6 +42,130 @@ function matrix_rep(g::DoubleExcitation)
+end
+Braket.chars(::Type{DoubleExcitationPlus}) = "G2+(ang)"
+Braket.qubit_count(::Type{DoubleExcitationPlus}) = 4
+Base.inv(g::DoubleExcitationPlus) = DoubleExcitationPlus(-g.angle[1])
```
What about gate exponentiation?

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -18,6 +42,130 @@ function matrix_rep(g::DoubleExcitation)
+Generate the matrix representation of the [DoubleExcitationMinus](https://docs.pennylane.ai/en/stable/code/api/pennylane.DoubleExcitationMinus.html) gate.
+
+This gate performs an SO(2) rotation in the subspace {|1100⟩, |0011⟩} with a phase-shift on other states:
+
```
Again with the different angle variables :)

Fe-r-oz replied:
sorry, I will use `ϕ `everywhere !

kshyatt-aws replied:
Either way, it's just good to be consistent!

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -31,6 +179,130 @@ function matrix_rep(g::SingleExcitation)
+
+This gate performs a rotation in adjacent fermionic modes under the Jordan-Wigner mapping, transforming states as follows:
+
+- |00⟩ ⟼ |00⟩
```
Angle variable names once again

Fe-r-oz commented:
@kshyatt-aws, A small request to kindly please assign/close the corresponding issue (14) as today the final day to assign/close issues! 

Thanks to your comments, I have significantly improved the PR!  Please feel free to review about your convenience. I am delighted to address any further comments.

Best,

Fe-r-oz commented:
Tested again,  Local Test Results: 

```
     Testing Running tests...
Test Summary:    |Time
Method ambiguity | None  0.0s
Test Summary:           | Pass  Total  Time
Unbound type parameters |    1      1  0.2s
Test Summary:     | Pass  Total  Time
Undefined exports |    1      1  0.0s
Test Summary:                              | Pass  Total  Time
Compare Project.toml and test/Project.toml |    1      1  0.2s
Test Summary:      | Pass  Total  Time
Stale dependencies |    1      1  6.4s
Test Summary: | Pass  Total  Time
Compat bounds |    4      4  0.7s
Test Summary: |Time
Piracy        | None  0.0s
Test Summary:    |Time
Persistent tasks | None  0.0s
Test Summary:   | Pass  Total     Time
BraketSimulator | 6800   6800  4m40.9s
     Testing BraketSimulator tests passed 
```

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -18,6 +43,132 @@ function matrix_rep(g::DoubleExcitation)
+Braket.chars(::Type{DoubleExcitationPlus}) = "G2+(ang)"
+Braket.qubit_count(::Type{DoubleExcitationPlus}) = 4
+Base.inv(g::DoubleExcitationPlus) = DoubleExcitationPlus(-g.angle[1])
+Base.:^(g::DoubleExcitationPlus, power::Integer) = power == -1 ? inv(g) : (power == 0 ? DoubleExcitationPlus((0.0,)) : (power < 0 ? inv(g^(-power)) : DoubleExcitationPlus((g.angle[1] * power,))))
```
if `power == 0`, why not return `Identity()`?

kshyatt-aws replied:
Also, better to use `iszero` for this check

Fe-r-oz replied:
Added `iszero `check. Identity() was not present in` BraketSimulator` so used `DoubleExcitationPlus((0.0,))`

kshyatt-aws replied:
It's `Braket.I()`, not `Identity()`, which is my mistake and why you couldn't find it :p

Fe-r-oz replied:
It's okay, I have changed to `Braket.I()` instead of ` DoubleExcitationPlus((0.0,))` and likewise.

kshyatt-aws reviewed (approved):
If CI doesn't pass, we'll merge this to a feature branch

kshyatt-aws commented:
My approval here should be enough but just in case feel free to ping me again if UH doesn't award the bounty and I'll reach out to them. Thanks for the contribution and multiple revisions!

Fe-r-oz commented:
All the tests are passing except the CI Documentation. Please have look at doctests (it seems to be a strange error it's showing). I am not sure what's causing that. Please let me know!

Btw, I have changed to `Braket.I() `instead of `DoubleExcitationPlus((0.0,))` and likewise locally. Let's make sure all the tests passes!

Fe-r-oz commented:
@kshyatt-aws, I fixed the CI doc error , but I when I committed and pushed the commit, a werid message which is as follows "@[Fe-r-oz](https://github.com/Fe-r-oz) dismissed [kshyatt-aws](https://github.com/kshyatt-aws)’s [stale review](https://github.com/amazon-braket/BraketSimulator.jl/pull/25#pullrequestreview-2128997668) via [81187a5](https://github.com/amazon-braket/BraketSimulator.jl/commit/81187a57f9d31178b4b44efece14490b8c06bc8a) [4 minutes ago](https://github.com/amazon-braket/BraketSimulator.jl/pull/25#event-13240386846)" is now appearing and approval disappeared. Kindly please run the CI (hope the CI passes for documentation as well), Very sorry for the inconvenience, I am not sure why this message appeare. I was just trying to fix the CI doc error as well also added `Braket.I() `as well.  🙏🏼 

```
julia> doctest(BraketSimulator)
┌ Warning: Unable to determine HTML(edit_link = ...) from remote HEAD branch, defaulting to "master".
│ Calling `git remote` failed with an exception. Set JULIA_DEBUG=Documenter to see the error.
│ Unless this is due to a configuration error, the relevant variable should be set explicitly.
└ @ Documenter ~/.julia/packages/Documenter/CJeWX/src/utilities/utilities.jl:640
[ Info: SetupBuildDirectory: setting up build directory.
[ Info: Doctest: running doctests.
[ Info: Skipped ExpandTemplates step (doctest only).
[ Info: Skipped CrossReferences step (doctest only).
[ Info: Skipped CheckDocument step (doctest only).
[ Info: Skipped Populate step (doctest only).
[ Info: Skipped RenderDocument step (doctest only).
Test Summary:             | Pass  Total   Time
Doctests: BraketSimulator |    1      1  12.4s
Test.DefaultTestSet("Doctests: BraketSimulator", Any[], 1, false, false, true, 1.718963335868262e9, 1.71896334831641e9, false, "/home/Transcendence/.julia/packages/Documenter/CJeWX/src/doctest.jl")

```

kshyatt-aws commented:
Heya @Fe-r-oz , we haven't forgotten about this PR! As you can see, we have a big feature branch in progress (which shouldn't affect this much) -- once that is merged, we can pull this into `main` as well.

kshyatt-aws commented on src/custom_gates.jl:
```diff
@@ -1,3 +1,35 @@
+Generate the matrix representation of the [DoubleExcitation](https://docs.pennylane.ai/en/stable/code/api/pennylane.DoubleExcitation.html) gate.
+
+This gate performs an SO(2) rotation in the subspace {|1100⟩, |0011⟩}, transforming the states as follows:
+
```
Can the LaTeX rendering be restored here?

Fe-r-oz replied:
Restored!

kshyatt-aws commented:
Thanks for the update! @Fe-r-oz, are you alright with me merging this into a feature branch that we can then merge once https://github.com/amazon-braket/BraketSimulator.jl/pull/36 is merged?

Fe-r-oz commented:
Sorry for that! 

I thought that the big feature branch got merged, that's why I thought to resolve the merge conflict. Please feel free to merge this after #36 if you like or otherwise. There are no hurries from my side.

kshyatt-aws commented:
No worries! Thanks for bearing with us on this :). I really want to get your (awesome!) changes in, so if we merge them into a feature branch it becomes 100% my responsibility.

kshyatt-aws commented:
Alright, I think we got all our blocking feature branches merged! @Fe-r-oz is it alright with you if I rebase this and add a few compatibility edits before running CI one last time and (hopefully) merging? If you want to handle that instead, go ahead.

Fe-r-oz commented:
Please go ahead with the rebase!

Fe-r-oz commented:
There were merge conflicts due to introduction of `mutable` and `qubitcount` for `SingleExcitation` and `DoubleExcitation` which has been resolved with latest master-merge. Indeed, these new gates needs these compatiblity edits as well!
